### PR TITLE
Decrease allocation rate in internal/pack

### PR DESCRIPTION
**Pull request by greatroar**
What is the purpose of this change? What does it change?
--------------------------------------------------------

Decrease the allocation rate in the packer by >40%:

    name             old time/op    new time/op    delta
    PackerManager-8     180ms ± 2%     180ms ± 1%     ~     (p=0.340 n=9+9)
    
    name             old speed      new speed      delta
    PackerManager-8   292MB/s ± 2%   293MB/s ± 1%     ~     (p=0.328 n=9+9)
    
    name             old alloc/op   new alloc/op   delta
    PackerManager-8    91.8kB ± 0%    72.6kB ± 0%  -20.89%  (p=0.000 n=8+8)
    
    name             old allocs/op  new allocs/op  delta
    PackerManager-8     1.41k ± 0%     0.78k ± 0%  -44.46%  (p=0.000 n=9+10)

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

No.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by MichaelEischer** [comment]
@greatroar Could you take a look at my review comments?

File: internal/repository/packer_manager.go [hunk 1]
```diff
@@ -57,7 +57,9 @@ func (r *packerManager) findPacker() (packer *Packer, err error) {
 	// search for a suitable packer
 	if len(r.packers) > 0 {
 		p := r.packers[0]
-		r.packers = r.packers[1:]
+		last := len(r.packers) - 1
+		r.packers[0], r.packers[last] = r.packers[last], nil
```
[the remark this answers is not in the document]
Comment (greatroar) [reply internal/repository/packer_manager.go:61]: Done.

File: internal/pack/pack.go [hunk 2]
```diff
@@ -277,41 +258,37 @@ func List(k *crypto.Key, rd io.ReaderAt, size int64) (entries []restic.Blob, err
...
-		pos += uint(e.Length)
+		pos += entry.Length
+		buf = buf[entrySize:]
 	}
 
 	return entries, nil
 }
+
+func parseHeaderEntry(p []byte) (b restic.Blob, err error) {
+	p = p[:entrySize]
```
[the remark this answers is not in the document]
Comment (greatroar) [reply internal/pack/pack.go:280]: I've added a check.

File: internal/pack/pack.go [hunk 3]
```diff
@@ -277,41 +258,37 @@ func List(k *crypto.Key, rd io.ReaderAt, size int64) (entries []restic.Blob, err
...
-
-		pos += uint(e.Length)
+		pos += entry.Length
+		buf = buf[entrySize:]
 	}
 
 	return entries, nil
 }
+
+func parseHeaderEntry(p []byte) (b restic.Blob, err error) {
```
[the remark this answers is not in the document]
Comment (greatroar) [reply internal/pack/pack.go:295]: Done.

**Comment by greatroar** [comment]
Rebased and addressed the comments, but now PackerManager gets slower:

```
name             old speed      new speed      delta
PackerManager-8   294MB/s ± 1%   292MB/s ± 1%   -0.77%  (p=0.009 n=10+10)
```

With the const version of entrySize, it's as fast as before.

**Review by fd0 — APPROVED** [review]
LGTM, thanks! I've fixed the error checks for the test code (ignoring errors because writing to a buffer should never error)

**Pull request merged**